fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/fluid/FluidVariantImpl.java

```diff
@@ -49,25 +48,27 @@ public static FluidVariant of(Fluid fluid, @Nullable NbtCompound nbt) {
 			}
 		}
 
-		if (nbt == null || fluid == Fluids.EMPTY) {
+		if (components.isEmpty() || fluid == Fluids.EMPTY) {
 			// Use the cached variant inside the fluid
 			return ((FluidVariantCache) fluid).fabric_getCachedFluidVariant();
 		} else {
 			// TODO explore caching fluid variants for non null tags.
-			return new FluidVariantImpl(fluid, nbt);
+			return new FluidVariantImpl(fluid, components);
 		}
 	}
 
-	private static final Logger LOGGER = LoggerFactory.getLogger("fabric-transfer-api-v1/fluid");
+	public static FluidVariant of(RegistryEntry<Fluid> fluid, ComponentChanges components) {
+		return of(fluid.value(), components);
+	}
 
 	private final Fluid fluid;
-	private final @Nullable NbtCompound nbt;
+	private final ComponentChanges components;
 	private final int hashCode;
 
-	public FluidVariantImpl(Fluid fluid, NbtCompound nbt) {
+	public FluidVariantImpl(Fluid fluid, ComponentChanges components) {
 		this.fluid = fluid;
-		this.nbt = nbt == null ? null : nbt.copy(); // defensive copy
```

Comment:
Are you sure that the defensive copy is not needed anymore?

Reply:
As far as I can tell ComponentChanges and all components are immutable.